src/vs/editor/common/languages.ts

```diff
@@ -176,11 +192,33 @@ export interface Hover {
  */
 export interface HoverProvider {
 	/**
-	 * Provide a hover for the given position and document. Multiple hovers at the same
+	 * Provide a hover for the given position, context and document. Multiple hovers at the same
 	 * position will be merged by the editor. A hover can have a range which defaults
 	 * to the word range at the position when omitted.
 	 */
-	provideHover(model: model.ITextModel, position: Position, token: CancellationToken): ProviderResult<Hover>;
+	provideHover(model: model.ITextModel, position: Position, token: CancellationToken, context?: HoverContext): ProviderResult<Hover>;
+}
+
+export interface HoverContext {
+	/**
+	 * Whether to increase or decrease the hover's verbosity
+	 */
+	action: HoverVerbosityAction;
+	/**
+	 * The previous hover for the same position
+	 */
+	hover: Hover;
```

Comment:
I'd make these fields optional - the context should be independent of verbosity.